### ISSUE TITLE
Create and Update Owner Status

### ISSUE THREAD

**Issue by joshdanielson** [loconomics/loconomics#279]
## Child Stories
- As  [Loconomics Staff](https://github.com/loconomics/loconomics/blob/master/docs/Personas/Sara%20the%20Service%20Professional.md) I need to **view** the  *owner status* of a user so I can tailor communications, marketing, and patronage calculations based on it.

### Test Cases 
1) Make sure owner statuses are clearly defined
2) Make sure owner statuses are updated when they change

### Front-end
- 
### JS
- 
### API
- 
### Database
- [x]  Add to user table: 
     - OwnerStatusID
     - OwnerAnniversaryDate (First date that they became an Owner...does not change after that)
- [x] Create OwnerStatus table:

| OwnerStatusID | OwnerStatusName | OwnerStatusDescription |
| --- | --- | --- |
| 0 | Not yet an owner | User is either a client or is a service professional that has not completed steps to become an owner but is either in trial or entered valid payment info |
| 1 | Trial ended | User's free trial has ended without them entering payment info |
| 2 | Active | Owner has completed two bookings, one listing, payment info, and owner acknowledgment |
| 3 | Inactive | Owner has at one time completed all required steps to becoming an owner but does not have at least one completed listing, or user fees are past due more than 30 days |
| 4 | Cancelled | User has cancelled their account |
| 5 | Suspended | User has been suspended by Loconomics Staff |

- [x] Create "UpateOwnerStatus" (see #279)
### Braintree
-  

## Notes
@IagoSRL commented on [Tue Jan 12 2016](https://github.com/joshdanielson/loconomics-internal/issues/847#issuecomment-170922097)

I did some changes on tables to adjust Primary and Foreign keys. I removed the language and country fields at OwnerStatus table in order to create the foreign key with [users] table. I will use a copy of the OwnerStatus descriptions in code, rather than fetch them from database (so future updates will need to update the App language code); is best this way for this table.

**Comment by IagoSRL** [comment]
The original description here lacks the need for a procedure at server that checks our bylaws rules against the user profile (2 bookings, ownerAcknocklegment, paymentPlan status), and updates the OwnerStatusID. 

Would work similar to our `TestAlert*` procs, that is called whenever a related change happens.

Specially, we may take into account if we have some status that is set manually by Loconomics stuff, so the automatic process must don't try to update the status on that cases.

**Comment by IagoSRL** [comment]
Part of commended work here was done per #334 and details.

In server code I created an enumeration that include a value 0 for OwnerStatus with name 'unset' without knowing the value was described here already (but not in database); I may rename it to 'Not yet an owner'.

The scheduled task/webhooks already updates the OwnerStatus when a subscription changes happens, not sure what additional work is expected here.

**Comment by joshdanielson** [comment]
@IagoSRL sorry this isn't clearer. So really this is in conjunction with #334 and this issue is really about a procedure checking for membership requirements like you mentioned above.

The only manual ones for now would be 4 or 5 (canceling the account or suspension by staff). We don't have a way for the user to cancel yet other than send us a customer service request (totally fine for now).

**Comment by IagoSRL** [comment]
@joshdanielson Right now, the code supports to set the OwnerStatus as cancelled when an update is received from Braintree as the subscription was cancelled, and suspended if was expired. This even means that if stuff member needs to cancel a plan, can do it just in Braintree and our database will get automatically updated (because of the Webhook/scheduled task --if everything works fine!).

Aside note: internally, I have a call to cancel the Braintree subscription, is a simple code used now to rollback the 'CreateSubscription' operation if something failed in the process. To offer this option to the user in the app just require to add the REST API call for this and implement the UI right now (since internal complexities are already implemented).

I will keep an eye on this issue to complete any detail remaining, but yes, the main feature is done at #334

**Comment by joshdanielson** [comment]
@IagoSRL nice work!! Let's leave the cancelation request manual for the user for now...will be good to ask why.

**Comment by IagoSRL** [comment]
I saw that the OwnerStatus zero exists at dev and live already, but was not in my local database, so I'm copying; something failed in the process of provide 'upgrade sql scripts' for data at some point here.
Some other records has changed descriptions, and 'In default' has gone. Copy of the table, with [A] being added and [D] deleted. @situatedbit check this in case you need to update your local DB too.

<pre>
[A] 0	Not yet an owner	User is either a client or is a service professional that has not completed steps to become an owner but is either in trial or entered valid payment info	2017-03-07 00:00:00.000	2017-03-07 00:00:00.000	True	jd
1	Trial ended	User's free trial has ended without them entering payment info	2015-12-01 00:00:00.000	2015-12-01 00:00:00.000	True	jd
2	Active	Owner has completed two bookings, one listing, payment info, and owner acknowledgment	2015-12-01 00:00:00.000	2015-12-01 00:00:00.000	True	jd
3	Inactive	Owner has at one time completed all required steps to becoming an owner but does not have at least one completed listing, or user fees are past due more than 30 days	2015-12-01 00:00:00.000	2015-12-01 00:00:00.000	True	jd
[D] 4	In default	Owner is past due on their payment.	2015-12-01 00:00:00.000	2015-12-01 00:00:00.000	True	jd
5	Cancelled	User has cancelled their account	2015-12-01 00:00:00.000	2015-12-01 00:00:00.000	True	jd
6	Suspended	User has been suspended by Loconomics Staff	2015-12-01 00:00:00.000	2015-12-01 00:00:00.000	True	jd
</pre>

@joshdanielson because 'In Default' has gone, I suppose I must remove support from code so I'm doing it. The value 'In trial' changed to 'Trial ended', there is no code right now to detect and change to that status right now (and rules may need to be adapted), so Josh clarify this please.

With that changes, logic to know if a user is in a valid state to have access to it's membership features, is just check the `OwnerStatus==OwnerStatus.Active //(2)` (plus check the chosen paymentPlan). This is important for #442 implementation.

**Comment by IagoSRL** [comment]
There was no task and description for the field `ownerAnniversaryDate`. I suppose must be filled with Date the first time user enters 'Active' OwnerStatus, so I just implement it that way.

About the 'trial ended' thing, some clarifications: right now, because of Braintree logic and notifications:
- when creating Braintree subscription, the payment method is checked, if is not correct even if it has a trial period Braintree aborts the operation (no subscription, no UserPaymentPlan).
- if is valid, subscription is set as 'active', and is charged later when the trial period ends
- if payment charge was success, subscription continues as 'active', if was unsuccessfully, enters due date then expiration (suspended in OwnerStatus).

Maybe we need another issue to clarify the logic about the 'trial period', when user enters that and conditions to set status 'trial ended' rather than inactive/suspended/cancelled.

**Comment by situatedbit** [comment]
I haven't tested any of the code on dev, but I read through the new code in ``LcRest/UserProfile.cs``. Functionally, I didn't see any issues. There are a few ways I think we can make the code easier to maintain the future:
1. Broadly, all of the owner status code could be moved out of UserProfile and into its own class. UserProfile would still have a status, but the logic can be elsewhere. A status can be its own object, and ``CheckOwnerStatus`` becomes ``public static OwnerStatus Get(int userID)`` or something similar.
2. We should remove any code we're not using, like ``MeetsBookingRequirement``. It's now in the repo and we can add it later if we need to use it
3. ``SetOwnerStatus`` sql: some of this SQL could be moved into the classes associated with their tables and called through methods on those classes, such as creating ``OwnerStatusHistory`` rows, setting the ``OwnerAnniversaryDate``, etc. 
4. Do you think we could start moving calls to the alert checks behind methods, too? Perhaps a method that executes ``TestAllUserAlerts``?
5. The sql in the ``Meets*`` methods could be executed behind methods in their respective classes. That way the owner status code doesn't need to know how the other parts of the code are storing their data: ``MeetsListingRequirement`` could call a method on ``LcRest.ServiceProfessionalService``, ``MeetsOwnerAcknowledgement`` could call a method on ``LcRest.OwnerAcknowledgement``, and so on.

**Comment by IagoSRL** [comment]
Thanks @situatedbit for the review.

1,3: I agree, had a similar feeling when doing it, but since didn't fit exactly like other models I didn't know exactly how name them (classes, methods; example, OwnerStatus.Get sounds like getting the stored value, but we are calculating one that is not saved at the point); similar with putting OwnerStatusHistory out of it, seems to need it's own class but at the same time is good to have the insertion in the same transaction (so commits all or nothing).
I will try to refactor some things per your advice, let me rest it a bit.

2: Per bylaws we will need to enable that again not too far from now, but get the point of unused code. Uhmm, maybe open an issue for re-enabling it and point to the removal commit?

4, 5: I tried with this, but not sure if will be better to use/maintain this way or in stored procs. Seems organized this way, and just one responsability (it checks the rules and give the result, but [UserAlert] keeps untouched --TestAlert* do both things).

**Comment by situatedbit** [comment]
## 1,3

For an owner status object, I was thinking an interface like this:

```c#
public int statusID
public OwnerStatus(int statusID)
public bool CanTransitionTo(OwnerStatus transitionStatus)
```
UserProfile could have a private instance of ``OwnerStatus`` so it isn't serialized through the API.

I haven't thought through the implications for the other lines where status is handled. Just an idea.

## 2

I think that's a good idea.

**Comment by IagoSRL** [comment]
@situatedbit I removed the bookings requirement an opened #484.

I refactored the code related to Owner status, is at is191 branch. I ended creating an Owner class rather than just OwnerStatus: I started trying this last but some details were weird and ended thinking that managing owner related in same class (userID, status and date) have more sense. Fits good too with the idea that the database [users] table must have less fields, good candidates to move to a 1-1 relationship table are the Owner* fields.
Still, some details don't *look* so good as expected but may be a step forward (I think) but let me know (we can still go back).
